Allow upload directly to S3 compatible storages

BoPeng opened the pull request:
@mbraak 

This is our PR that allows upload directly to S3. We are still working on some problems (e.g. upload sometimes fail without clear reason, dropping of identical files does not replace existing) but we feel that it is time to listen to your opinion on it.

The PR roughly works like this

1. User defines all S3 related variables as required by `django-storage`.
2. User specify `s3_upload_dir` in `View`, which adds a hidden field to the form, similar to the `placeholder` field.
3. When user upload, the files are uploaded to S3 directly. We borrowed code from `uppy/s3` for the JS part of the implementation.
4. Upon success, the information is written to the hidden field.
5. When the form is saved, the information is extracted and is used to create objects derived from `S3Boto3StorageFile`.
6. Users use `is_s3direct` to judge the type of the object, which could be removed since there is no way to mix regular local upload and s3 direct upload. Reading from the object will download the file from S3.

We added two test cases but we are not sure how to run the test in travis.

mbraak commented:
I will have a look

mbraak commented:
Thanks for the pr! This is a very useful feature if you want to upload large files.

I'm still setting up a test environment (using localstack). In the meantime, I did get errors / warnings from `eslint` and `typescript`:

```
cd frontend
yarn run lint
yarn run tsc
```

BoPeng commented:
@jma7 wrote the JS S3 part and I simply used `webpack --mode production` to get it running.

@jma7 Could you fix the warning and errors reported by `lint` and `tsc`? I squashed the commits to create this PR so you should work on the `s3direct` branch. Thanks.

BoPeng commented on django_file_form/views/s3multipart.py:
```diff
@@ -0,0 +1,185 @@
+            json_body = json.loads(request.body)
+            fileName = json_body["filename"]
+            contentType = json_body["contentType"]
+            key = cls.file_form_upload_dir + "/" + fileName
```
@jma7 Is it possible to add an unique ID between `FILE_FORM_UPLOAD_DIR` and `fileName`? Right now if multiple users upload to the same S3 account, the files will clash. It is better to add an per-form or per-user ID here. Perhaps we should even add another layer of ID and allow users to specify a directory in addition to `FILE_FORM_UPLOAD_DIR` (e.g. multiple forms inside an application). I suppose in this case the `direct_to_s3` variable in `View` could be used.

In this end we could have

```
BUCKET_NAME/UNIQUE_ID/USER_FOLDER/FILENAME
```

BoPeng commented on django_file_form/urls.py:
```diff
@@ -1,8 +1,14 @@
+    path('s3upload/<uploadId>/', s3multipart.getUploadedParts, name='getUploadedParts'),
+    path('s3upload/<uploadId>/<int:partNumber>', s3multipart.signPartUpload, name='signPartUpload'),
+    path('s3upload/<uploadId>/complete', s3multipart.completeMultipartUpload, name='completeMultipartUpload'),
+    path('s3upload/<uploadId>/', s3multipart.abortMultipartUpload, name='abortMultipartUpload'),
```
@jma7 Why is this URL the same as `getUploadedParts`, will this cause any problem?

BoPeng commented on frontend/src/upload_file.ts:
```diff
@@ -1,27 +1,30 @@
-  placeholder?: boolean;
+  placeholder?: boolean | undefined;
   size: number;
+  url?: string;
```
Not quite sure about this one. It should still be in `UploadedFile`, right?

BoPeng commented on frontend/src/upload_file.ts:
```diff
@@ -281,7 +305,8 @@ class UploadFile {
     this.renderer.deleteFile(uploadIndex);
-    delete this.uploads[uploadIndex];
+    // delete this.uploads[uploadIndex];
+    this.uploads.splice(uploadIndex, 1)
```
@jma7 This looks suspicious. Please double check.

jma7 replied:
Fixed.

BoPeng commented on testproject/django_file_form_example/forms.py:
```diff
@@ -63,6 +63,23 @@ def save(self):
+        )
+        for f in self.cleaned_data['input_file']:
+            assert f.is_s3direct
+            # download from S23
```
typo

BoPeng commented on testproject/testproject/settings_default.py:
```diff
@@ -23,7 +23,8 @@ def create_media_paths(media_root):
     default=dict(
-        ENGINE='django.db.backends.postgresql',
+        # ENGINE='django.db.backends.postgresql',
+        ENGINE='django.db.backends.sqlite3',
```
This should be reverted.

BoPeng commented on testproject/testproject/settings_default.py:
```diff
@@ -89,4 +90,18 @@ def create_media_paths(media_root):
+CSP_FONT_SRC = ("'self'")
+CSP_IMG_SRC = ("'self'",)
+CSP_CONNECT_SRC = ("'self'", os.environ.get("AWS_S3_ENDPOINT_URL", "") or os.environ.get("DJANGO_AWS_S3_ENDPOINT_URL", "") )
+
```
These should be documented in `README.md`.

mbraak commented on testproject/requirements.txt:
```diff
@@ -1,2 +1,4 @@
 pytz==2020.1
 django-csp==3.6
+boto3=1.14.20
```
I got an error trying to install the `requirements.txt`. I think `==` is expected.

BoPeng replied:
Will fix soon.

mbraak commented on testproject/django_file_form_example/templates/example_form.html:
```diff
@@ -25,6 +26,11 @@
             {{ form.title }}
         </div>
+
+        {% if direct_to_s3 %}
```
Is it correct that the upload method is set in this example project? I would expect it to be set in the `django-file-form` app.

Would it be possible to put the upload method in a `data` attribute, instead of an input? Perhaps it can be put on `dff-container`, like `data-files`. See `upload_widget.htm`.
The reason that I ask this is that an input could cause a conflict; somebody could add an `uploadMethod` input to a form.

BoPeng replied:
I agree that `example_form.html` is the wrong place for this piece of code since this requires users to add the HTML code to their own forms. `upload_widget.htm` is a much better place but as I commented below, I do not have a good way to pass such as information to this widget.

BoPeng replied:
This is now done by adding a hidden field.

mbraak replied:
I see that the `s3_upload_dir` is now added in `django_file_form`. That's good.

mbraak commented:
After some CORS configuration I was able to upload a file using S3 as a temporary storage. And I could even save the file in `ExampleFile` to S3 using `S3Boto3Storage`.

I added two comments to the code.

BoPeng commented:
@jma7 @mbraak 

Besides some minor code verification (@jma7 please check my comments), I believe function-wise there is only one problem remains.

Right now everything is written as `{FILE_FORM_UPLOAD_DIR}/{filename}` which is impractical for a multi-user environment. I have changed `direct_to_s3` to `s3_upload_dir` so that the files will be uploaded as

```
{FILE_FORM_UPLOAD_DIR}/{s3_upload_dir}/{filename}
```
Here the developers can choose to create a per-user ID, or per-form ID, even per-session ID so that files will be uploaded to different directories under S3. Currently `s3_upload_dir` is passed through `<input>` and is a flag only.

Now, I am having difficulties in 1) passing the information to the frontend, 2) telling the AJAX uploader to use this directory, and 3) using it in `S3Boto3StorageFile` and I will need your help.

1. It is possible to set this as a `data` attribute to `dff-uploader` or `dff-container`. However, the template is expanded in

https://github.com/mbraak/django-file-form/blob/0a1e50589e30deb23833b8c2f2c9739b1d7cf68b/django_file_form/widgets.py#L45-L57

and I do not know how to pass something like `data-s3-upload-dir={{ s3_upload_dir }}` to it. Note that this piece of HTML is expanded before the rendering of the template so `s3_upload_dir` is not available even though it is [in `context`](https://github.com/bioworkflows/django-file-form/blob/dbb02ae8467bbebebfc9a406c3b6c8745b34d6bf/testproject/django_file_form_example/views.py#L27).

2. Now, if the information is passed to the frontend, how to use it during upload to S3? I can imagine using it in the AJAX side, or on the python side when [`key` is defined](https://github.com/bioworkflows/django-file-form/blob/s3direct/django_file_form/views/s3multipart.py#L46). However, currently the function is a classmethod and there is no way to pass this information over there.

3. Then if the file is uploaded correctly to S3, we will have to use it in `S3Boto3StorageFile` https://github.com/bioworkflows/django-file-form/blob/dbb02ae8467bbebebfc9a406c3b6c8745b34d6bf/django_file_form/models.py#L127-L129 which is probably an easier job when the first two problems are resolved.

BoPeng commented:
https://github.com/mbraak/django-file-form/pull/330/commits/2895ea8645022c688e19deab807b25ab9fc26728

This patch solves 1, namely passing `s3_upload_dir` to frontend through a hidden field, the same way as `form_id` etc.

@lma7, it is your turn to work on 2.

mbraak commented:
I saw one thing that might not be correct. If I upload the same filename in two browsers, then `temp_uploads` contains one file. Perhaps the file name should be a uuid.

BoPeng commented:
> I saw one thing that might not be correct. If I upload the same filename in two browsers, then temp_uploads contains one file. Perhaps the file name should be a uuid.

If the same user uploads from the same form on two browsers or machines, the user should expect the same file. However, files should be different from two users or two forms and this is exactly why I have been working on the use of per-user or per-session directory `s3_upload_dir`.

Another possible solution is to separate file content and filename. That is to say, we return filename to the backend while saving content to S3 with SHA1 filename derived from content. This will make sure that files with different content will have different filename on S3, and will actually save space if multiple files with the same content are uploaded. Although this could be an interesting solution, I suspect users do not expect change of filename on S3.

mbraak commented:
I believe the Django `FileSystemStorage` checks if a filename is unique. If the filename is not unique it adds a random string. Would it be possible to use something like that?

BoPeng commented:
> I believe the Django FileSystemStorage checks if a filename is unique. If the filename is not unique it adds a random string. Would it be possible to use something like that?

Yes, this is done at the django side with [`get_available_name`](https://docs.djangoproject.com/en/3.1/ref/files/storage/#django.core.files.storage.Storage.get_available_name)

We are a little different here because we are not saving the content to S3, rather construct an `S3Boto3StorageFile` from existing S2 content. I think similar thing at be done at 

https://github.com/bioworkflows/django-file-form/blob/e543098f73fc86b20745ae76ed3c45c464679cdc/django_file_form/views/s3multipart.py#L46

where the client can check if `key` exists and add a random string to `key` if needed.

However, for my particular application filenames are significant so I will need to get hold of the original filename, and I would something like

```
f.original_filename
```
if `f.file.name` can be different from what users uploaded.

BoPeng commented:
Done. Random bits are added to filename if object already exists, and `f.original_name` is added to get the original filename. I believe that we have addressed all issues with this PR.

mbraak commented:
I can have a look at the lint errors in `s3_uploader.ts`.

mbraak commented on django_file_form/views/s3multipart.py:
```diff
@@ -0,0 +1,184 @@
+        return name
+
+    @classmethod
+    def createMultipartUpload(cls, request):
```
Could you use snake case for this (`create_multipart_upload`)? Also applies to `getPartsOrAbortUpload` and other methods.

BoPeng replied:
Will do later but you could also push to our branch if it is easier for you.

BoPeng replied:
Done. Hopefully did not introduce any error.

mbraak commented:
If I open the testproject on http://localhost:8000/multiple I get an error message in the browser console:

```
util.ts:35 Cannot find input with name 'example-s3_upload_dir'
```

mbraak commented:
Somehow the following does not work anymore:

* Run testproject and go to http://localhost:8000/multiple
* Upload a file and leave the title field empty
* Submit
* -> message 'Title field is required'
* -> Uploaded file is still displayed (this is correct)
* Delete the file
* Submit
* -> The uploaded file is back
* NB: this does work on `master`

BoPeng commented:
Perhaps the style change and merge from master broke something. I will have a look tomorrow.

mbraak commented:
I fixed the eslint errors in `s3_uploader.ts`.

BoPeng commented:
> Somehow the following does not work anymore:

The problem is caused by the mixed use of `.placeholder`, which is `false` for both `s3` and regular upload. I am submitting a patch that uses `true` for `placeholder`, `false` for `S3` and `null` for regular, but it might make more sense to use something like `type = 'placeholder | s3direct | uploaded` later, especially if someone would like to add other upload types such as upload from / to dropbox.

BoPeng commented:
@mbraak I have fixed the bug. I noticed that

1. upload 3 files
2. re-upload the second file
3. the re-uploaded file is moved to the last, with index 0.
4. delete the re-uploaded file  fails

BoPeng commented:
Did some stress test by uploading 1000+ files simultaneously and about 10 uplods immediately failed with messages like

![image](https://user-images.githubusercontent.com/9889312/90844807-63293d00-e32a-11ea-8250-3e0f10ac6127.png)

jma7 commented:
> Did some stress test by uploading 1000+ files simultaneously and about 10 uplods immediately failed with messages like
> 
> ![image](https://user-images.githubusercontent.com/9889312/90844807-63293d00-e32a-11ea-8250-3e0f10ac6127.png)
> 
> Edit: According to [this article](https://stackoverflow.com/questions/24122506/neterr-insufficient-resources-error-when-adding-numerous-img-elements-to-dom), this could be a problem with chrome, not our app.

Have you tried in FireFox?

BoPeng commented:
> Have you tried in FireFox?

Just tried. No, firefox does not produce any error, but it becomes very slow at the end, to a point that firefox asked me if I would like to stop the script. 

I think our design is not particular friendly to a large number of files. For example, we need to **collect information of all uploaded files, and `json.dump` to the hidden field 1000+ times**. 

Anyway, chrome just finished uploading a 12G file and half way through the upload a 25G file, this was not at all possible before so I am very happy now.

BoPeng commented:
> util.ts:35 Cannot find input with name 'example-s3_upload_dir'

This is because the call to 

```
  const s3UploadDir = getInputValue("s3_upload_dir");
```

I think it is safe to remove this `error` message.

mbraak commented:
Merged
